core/src/main/scala/cats/data/Chain.scala

```diff
@@ -256,6 +256,93 @@ sealed abstract class Chain[+A] extends ChainCompat[A] {
     result
   }
 
+  /**
+   * take a certain amount of items from the front of the Chain
+   */
+  final def take(count: Int): Chain[A] = {
+    // invariant count >= 1
+    @tailrec
+    def go(lhs: Chain[A], count: Int, arg: Chain[A], rhs: Chain[A]): Chain[A] =
+      arg match {
+        case Wrap(seq) =>
+          if (count == 1) {
+            lhs.append(seq.head)
+          } else {
+            // count > 1
+            val taken = seq.take(count)
+            // we may have not takeped all of count
+            val newCount = count - taken.length
+            val wrapped = Wrap(taken)
+            // this is more efficient than using concat
+            val newLhs = if (lhs.isEmpty) wrapped else Append(lhs, wrapped)
+            if (newCount > 0) {
+              // we have to keep takeping on the rhs
+              go(newLhs, newCount, rhs, Chain.nil)
```

Comment:
Shouldn't we eagerly exit with the `newLhs` if `rhs.isEmpty == true`?

Reply:
I don't think so, we have to check for empty anyway on the next call call of the loop. If we also check here we are checking twice. If we could ensure we never see Empty in the loop it would be worth it, but it would mean we would we have uglier code: we don't have a total match, we have one arm (Empty) that is unreachable. I don't think the performance win would be there and it would complicate the code.

What do you think?

Reply:
Okay, actually, I took this suggestion. I remembered we have `NonEmpty` so we can recurse on that instead, that naturally pushes the check up to the places you suggest.

I still don't think it's any faster (since the Empty was the final check and always a terminal condition), but at least it's nice to be clear in the recursion that we are only operating on NonEmptyChain.